File: macos/Protocol/FileTransfer.swift
import Foundation

public struct OutgoingFile {
    public var meta: FileMeta
    public var open: () throws -> InputStream
    public init(meta: FileMeta, open: @escaping () throws -> InputStream) {
        self.meta = meta
        self.open = open
    }
}

public protocol ReceiveSink: AnyObject {
    var bytesWritten: Int64 { get }
    func write(_ data: Data) throws
    func commit(expectedSha256: String) throws
    func abort()
}

public protocol ReceiveSinkFactory {
    func hasSpace(bytes: Int64) -> Bool
    func open(file: FileMeta, offset: Int64) throws -> ReceiveSink
}

public struct TransferProgress: Equatable {
    public var transferId: String
    public var filesCompleted: Int
    public var filesTotal: Int
    public var bytesTransferred: Int64
    public var bytesTotal: Int64
    public var currentName: String
    public var currentBytes: Int64
    public var currentSize: Int64
    public var bytesPerSecond: Double
    public var etaSeconds: Int64?
    public var state: TransferState

    public init(transferId: String, filesCompleted: Int, filesTotal: Int, bytesTransferred: Int64, bytesTotal: Int64, currentName: String, currentBytes: Int64, currentSize: Int64, bytesPerSecond: Double, etaSeconds: Int64?, state: TransferState) {
        self.transferId = transferId
        self.filesCompleted = filesCompleted
        self.filesTotal = filesTotal
        self.bytesTransferred = bytesTransferred
        self.bytesTotal = bytesTotal
        self.currentName = currentName
        self.currentBytes = currentBytes
        self.currentSize = currentSize
        self.bytesPerSecond = bytesPerSecond
        self.etaSeconds = etaSeconds
        self.state = state
    }
}

public final class FileTransfer {
    private let session: ProtocolSession
    public var cancelled = false

    public init(session: ProtocolSession) {
        self.session = session
    }

    public func send(request: TransferRequest, files: [OutgoingFile], onProgress: (TransferProgress) -> Void) throws {
        try session.send(makeEnvelope(type: .transferRequest, payload: request.json))
        let response = try session.receiveControl()
        if response.type == MessageType.transferRejected.rawValue {
            let reason = response.payload["reason"]?.string ?? ""
            throw ProtocolError(reason == "INSUFFICIENT_STORAGE" ? .diskFull : .cancelled, reason)
        }
        if response.type != MessageType.transferAccepted.rawValue {
            throw ProtocolError(.protocolViolation, "expected TRANSFER_ACCEPTED")
        }
        let skipIds: Set<String>
        if case .array(let values) = response.payload["skipFileIds"] {
            skipIds = Set(values.compactMap { $0.string })
        } else {
            skipIds = []
        }
        let outgoing = files.filter { !skipIds.contains($0.meta.fileId) }
        let skippedBytes = files.filter { skipIds.contains($0.meta.fileId) }.reduce(Int64(0)) { $0 + $1.meta.size }
        let estimator = SpeedEstimator()
        var overall: Int64 = skippedBytes
        let alreadyDone = files.count - outgoing.count
        for (index, file) in outgoing.enumerated() {
            try throwIfCancelled(request.transferId)
            try sendOne(transferId: request.transferId, file: file, index: alreadyDone + index, totalFiles: files.count, overallBefore: overall, totalBytes: request.totalBytes, estimator: estimator, onProgress: onProgress)
            overall += file.meta.size
        }
        try session.send(makeEnvelope(type: .transferComplete, payload: ["transferId": .string(request.transferId)]))
        onProgress(TransferProgress(transferId: request.transferId, filesCompleted: files.count, filesTotal: files.count, bytesTransferred: request.totalBytes, bytesTotal: request.totalBytes, currentName: "", currentBytes: 0, currentSize: 0, bytesPerSecond: estimator.bytesPerSecond, etaSeconds: 0, state: .completed))
    }

    public func receive(sinkFactory: ReceiveSinkFactory, accept: (TransferRequest) -> ReceiveDecision, onProgress: (TransferProgress) -> Void) throws -> TransferRequest {
        let envelope = try session.receiveControl()
        guard envelope.type == MessageType.transferRequest.rawValue else {
            throw ProtocolError(.protocolViolation, "expected TRANSFER_REQUEST")
        }
        let request = try TransferRequest.from(envelope.payload)
        let decision = accept(request)
        let skipIds = Set(decision.skipFileIds)
        let needed = decision.neededBytes ?? request.files.filter { !skipIds.contains($0.fileId) }.reduce(Int64(0)) { $0 + $1.size }
        if !decision.accepted {
            try session.send(makeEnvelope(type: .transferRejected, payload: ["transferId": .string(request.transferId), "reason": .string("USER_DECLINED")]))
            throw ProtocolError(.cancelled, "user declined")
        }
        if !sinkFactory.hasSpace(bytes: needed) {
            try session.send(makeEnvelope(type: .transferRejected, payload: ["transferId": .string(request.transferId), "reason": .string("INSUFFICIENT_STORAGE")]))
            throw ProtocolError(.diskFull, "not enough storage")
        }
        try session.send(makeEnvelope(type: .transferAccepted, payload: [
            "transferId": .string(request.transferId),
            "skipFileIds": .array(decision.skipFileIds.map { .string($0) }),
        ]))
        let estimator = SpeedEstimator()
        var overall: Int64 = request.files.filter { skipIds.contains($0.fileId) }.reduce(0) { $0 + $1.size }
        var completed = skipIds.count
        while true {
            try throwIfCancelled(request.transferId)
            switch try session.receive() {
            case .control(let data):
                let control = try ProtocolJSON.decode(data)
                switch MessageType(rawValue: control.type) {
                case .fileStart:
                    let start = try FileStart.from(control.payload)
                    overall = try receiveOne(request: request, start: start, sinkFactory: sinkFactory, completed: completed, overallBefore: overall, estimator: estimator, onProgress: onProgress)
                    completed += 1
                case .transferComplete:
                    onProgress(TransferProgress(transferId: request.transferId, filesCompleted: completed, filesTotal: request.files.count, bytesTransferred: request.totalBytes, bytesTotal: request.totalBytes, currentName: "", currentBytes: 0, currentSize: 0, bytesPerSecond: estimator.bytesPerSecond, etaSeconds: 0, state: .completed))
                    return request
                case .transferCancelled:
                    throw ProtocolError(.cancelled, "peer cancelled")
                default:
                    throw ProtocolError(.protocolViolation, "unexpected \(control.type)")
                }
            case .binary:
                throw ProtocolError(.protocolViolation, "binary without FILE_START")
            }
        }
    }

    public func cancel(transferId: String) {
        cancelled = true
        try? session.send(makeEnvelope(type: .transferCancelled, payload: ["transferId": .string(transferId), "reason": .string("USER_CANCELLED")]))
    }

    private func sendOne(transferId: String, file: OutgoingFile, index: Int, totalFiles: Int, overallBefore: Int64, totalBytes: Int64, estimator: SpeedEstimator, onProgress: (TransferProgress) -> Void) throws {
        try session.send(makeEnvelope(type: .fileStart, payload: [
            "transferId": .string(transferId),
            "fileId": .string(file.meta.fileId.lowercased()),
            "name": .string(file.meta.name),
            "size": .int(file.meta.size),
            "mimeType": .string(file.meta.mimeType),
            "offset": .int(0),
        ]))
        let digest = IncrementalSHA256()
        var sent: Int64 = 0
        let stream = try file.open()
        stream.open()
        defer { stream.close() }
        var buffer = [UInt8](repeating: 0, count: ProtocolConstants.chunkSize)
        while true {
            try throwIfCancelled(transferId)
            let n = stream.read(&buffer, maxLength: buffer.count)
            if n < 0 { throw ProtocolError(.fileUnavailable, "read failed") }
            if n == 0 { break }
            let chunk = Data(buffer[0..<n])
            digest.update(chunk)
            try session.sendBinary(fileId: UUID(uuidString: file.meta.fileId) ?? UUID(), offset: UInt64(sent), data: chunk)
            sent += Int64(n)
            let overall = overallBefore + sent
            let speed = estimator.onProgress(bytesTransferred: overall, nowMs: nowMs())
            onProgress(TransferProgress(transferId: transferId, filesCompleted: index, filesTotal: totalFiles, bytesTransferred: overall, bytesTotal: totalBytes, currentName: file.meta.name, currentBytes: sent, currentSize: file.meta.size, bytesPerSecond: speed, etaSeconds: estimator.etaSeconds(remaining: totalBytes - overall), state: .transferring))
        }
        if sent != file.meta.size {
            throw ProtocolError(.fileUnavailable, "size changed while sending")
        }
        try session.send(makeEnvelope(type: .fileComplete, payload: [
            "transferId": .string(transferId),
            "fileId": .string(file.meta.fileId.lowercased()),
            "bytes": .int(sent),
            "sha256": .string(digest.hex()),
        ]))
    }

    private func receiveOne(request: TransferRequest, start: FileStart, sinkFactory: ReceiveSinkFactory, completed: Int, overallBefore: Int64, estimator: SpeedEstimator, onProgress: (TransferProgress) -> Void) throws -> Int64 {
        let relative = try FilenameConflict.sanitizeRelativePath(start.name)
        let meta = FileMeta(fileId: start.fileId, name: (relative as NSString).lastPathComponent, size: start.size, mimeType: start.mimeType, relativePath: relative)
        let sink = try sinkFactory.open(file: meta, offset: start.offset)
        let digest = IncrementalSHA256()
        do {
            while sink.bytesWritten < start.size {
                try throwIfCancelled(request.transferId)
                switch try session.receive() {
                case .binary(let fileId, _, let data):
                    if fileId.uuidString.lowercased() != start.fileId.lowercased() {
                        throw ProtocolError(.protocolViolation, "unexpected file id")
                    }
                    digest.update(data)
                    try sink.write(data)
                    let overall = overallBefore + sink.bytesWritten
                    let speed = estimator.onProgress(bytesTransferred: overall, nowMs: nowMs())
                    onProgress(TransferProgress(transferId: request.transferId, filesCompleted: completed, filesTotal: request.files.count, bytesTransferred: overall, bytesTotal: request.totalBytes, currentName: meta.name, currentBytes: sink.bytesWritten, currentSize: start.size, bytesPerSecond: speed, etaSeconds: estimator.etaSeconds(remaining: request.totalBytes - overall), state: .transferring))
                case .control(let data):
                    let control = try ProtocolJSON.decode(data)
                    if control.type == MessageType.fileComplete.rawValue {
                        let sha = control.payload["sha256"]?.string ?? ""
                        if !Checksums.equalsHex(sha, digest.hex()) {
                            throw ProtocolError(.checksumMismatch, "hash mismatch")
                        }
                        try sink.commit(expectedSha256: sha)
                        return overallBefore + sink.bytesWritten
                    }
                    if control.type == MessageType.transferCancelled.rawValue {
                        sink.abort()
                        throw ProtocolError(.cancelled, "peer cancelled")
                    }
                    throw ProtocolError(.protocolViolation, "unexpected \(control.type)")
                }
            }
            let complete = try session.receiveControl()
            let sha = complete.payload["sha256"]?.string ?? ""
            if !Checksums.equalsHex(sha, digest.hex()) {
                throw ProtocolError(.checksumMismatch, "hash mismatch")
            }
            try sink.commit(expectedSha256: sha)
            return overallBefore + sink.bytesWritten
        } catch {
            sink.abort()
            throw error
        }
    }

    private func throwIfCancelled(_ transferId: String) throws {
        if cancelled {
            try session.send(makeEnvelope(type: .transferCancelled, payload: ["transferId": .string(transferId), "reason": .string("USER_CANCELLED")]))
            throw ProtocolError(.cancelled, "cancelled")
        }
    }

    private func nowMs() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }
}

public struct FileStart {
    public var transferId: String
    public var fileId: String
    public var name: String
    public var size: Int64
    public var mimeType: String
    public var offset: Int64
    public init(transferId: String, fileId: String, name: String, size: Int64, mimeType: String, offset: Int64) {
        self.transferId = transferId
        self.fileId = fileId
        self.name = name
        self.size = size
        self.mimeType = mimeType
        self.offset = offset
    }
    public static func from(_ payload: [String: JSONValue]) throws -> FileStart {
        guard let transferId = payload["transferId"]?.string,
              let fileId = payload["fileId"]?.string,
              let name = payload["name"]?.string,
              let size = payload["size"]?.int else {
            throw ProtocolError(.protocolViolation, "bad FILE_START")
        }
        return FileStart(transferId: transferId, fileId: fileId, name: name, size: size, mimeType: payload["mimeType"]?.string ?? "application/octet-stream", offset: payload["offset"]?.int ?? 0)
    }
}

File: android/protocol/src/main/kotlin/com/honor/share/protocol/FileTransfer.kt
package com.honor.share.protocol

import java.io.InputStream
import java.io.OutputStream
import java.util.UUID
import java.util.concurrent.atomic.AtomicBoolean

data class OutgoingFile(
    val meta: FileMeta,
    val open: () -> InputStream,
)

interface ReceiveSink {
    val bytesWritten: Long
    fun write(data: ByteArray)
    fun commit(expectedSha256: String)
    fun abort()
}

interface ReceiveSinkFactory {
    fun open(file: FileMeta, offset: Long): ReceiveSink
    fun hasSpace(bytes: Long): Boolean
}

data class TransferProgress(
    val transferId: String,
    val filesCompleted: Int,
    val filesTotal: Int,
    val bytesTransferred: Long,
    val bytesTotal: Long,
    val currentName: String,
    val currentBytes: Long,
    val currentSize: Long,
    val bytesPerSecond: Double,
    val etaSeconds: Long?,
    val state: TransferState,
)

data class ReceiveDecision(
    val accepted: Boolean,
    val skipFileIds: List<String> = emptyList(),
    val neededBytes: Long? = null,
)

class FileTransfer(
    private val session: ProtocolSession,
    private val cancelled: AtomicBoolean = AtomicBoolean(false),
) {
    fun send(
        request: TransferRequestPayload,
        files: List<OutgoingFile>,
        onProgress: (TransferProgress) -> Unit,
    ) {
        session.send(session.typed(MessageType.TRANSFER_REQUEST, request))
        val response = session.receiveControl()
        val skipIds: Set<String> = when (response.messageType()) {
            MessageType.TRANSFER_ACCEPTED -> {
                runCatching { ProtocolJson.payload<TransferAcceptedPayload>(response).skipFileIds }
                    .getOrDefault(emptyList())
                    .toSet()
            }
            MessageType.TRANSFER_REJECTED -> {
                val reason = ProtocolJson.payload<TransferRejectedPayload>(response).reason
                val code = if (reason == "INSUFFICIENT_STORAGE") ErrorCode.DISK_FULL else ErrorCode.CANCELLED
                throw ProtocolException(code, reason)
            }
            MessageType.ERROR -> {
                val error = ProtocolJson.payload<ErrorPayload>(response)
                throw ProtocolException(ErrorCode.fromWire(error.code), error.code)
            }
            else -> throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "expected TRANSFER_ACCEPTED")
        }
        val outgoing = files.filter { it.meta.fileId !in skipIds }
        val skippedBytes = files.filter { it.meta.fileId in skipIds }.sumOf { it.meta.size }
        val estimator = SpeedEstimator()
        var overall = skippedBytes
        val alreadyDone = files.size - outgoing.size
        outgoing.forEachIndexed { index, file ->
            throwIfCancelled(request.transferId)
            sendOne(
                request.transferId,
                file,
                alreadyDone + index,
                files.size,
                overall,
                request.totalBytes,
                estimator,
                onProgress,
            )
            overall += file.meta.size
        }
        session.send(session.typed(MessageType.TRANSFER_COMPLETE, TransferIdPayload(request.transferId)))
        onProgress(
            TransferProgress(
                transferId = request.transferId,
                filesCompleted = files.size,
                filesTotal = files.size,
                bytesTransferred = request.totalBytes,
                bytesTotal = request.totalBytes,
                currentName = "",
                currentBytes = 0,
                currentSize = 0,
                bytesPerSecond = estimator.bytesPerSecond(),
                etaSeconds = 0,
                state = TransferState.COMPLETED,
            ),
        )
    }

    fun receive(
        sinkFactory: ReceiveSinkFactory,
        accept: (TransferRequestPayload) -> ReceiveDecision,
        onProgress: (TransferProgress) -> Unit,
    ): TransferRequestPayload {
        val envelope = session.receiveControl()
        if (envelope.messageType() != MessageType.TRANSFER_REQUEST) {
            throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "expected TRANSFER_REQUEST got ${envelope.type}")
        }
        val request = ProtocolJson.payload<TransferRequestPayload>(envelope)
        if (request.files.size > ProtocolConstants.MAX_FILES_PER_TRANSFER) {
            session.send(
                session.typed(
                    MessageType.TRANSFER_REJECTED,
                    TransferRejectedPayload(request.transferId, "TOO_MANY_FILES"),
                ),
            )
            throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "too many files")
        }
        val decision = accept(request)
        val skipIds = decision.skipFileIds.toSet()
        val needed = decision.neededBytes
            ?: request.files.filter { it.fileId !in skipIds }.sumOf { it.size }
        if (!decision.accepted) {
            session.send(
                session.typed(
                    MessageType.TRANSFER_REJECTED,
                    TransferRejectedPayload(request.transferId, "USER_DECLINED"),
                ),
            )
            throw ProtocolException(ErrorCode.CANCELLED, "user declined")
        }
        if (!sinkFactory.hasSpace(needed)) {
            session.send(
                session.typed(
                    MessageType.TRANSFER_REJECTED,
                    TransferRejectedPayload(request.transferId, "INSUFFICIENT_STORAGE"),
                ),
            )
            throw ProtocolException(ErrorCode.DISK_FULL, "not enough storage")
        }
        session.send(
            session.typed(
                MessageType.TRANSFER_ACCEPTED,
                TransferAcceptedPayload(request.transferId, decision.skipFileIds),
            ),
        )
        val estimator = SpeedEstimator()
        val skippedBytes = request.files.filter { it.fileId in skipIds }.sumOf { it.size }
        var overall = skippedBytes
        var completed = skipIds.size
        while (true) {
            throwIfCancelled(request.transferId)
            val next = session.receive()
            when (next) {
                is Frame.Control -> {
                    val control = ProtocolJson.parse(next.json)
                    when (control.messageType()) {
                        MessageType.FILE_START -> {
                            val start = ProtocolJson.payload<FileStartPayload>(control)
                            overall = receiveOne(
                                request,
                                start,
                                sinkFactory,
                                completed,
                                overall,
                                estimator,
                                onProgress,
                            )
                            completed++
                        }
                        MessageType.TRANSFER_COMPLETE -> {
                            onProgress(
                                TransferProgress(
                                    transferId = request.transferId,
                                    filesCompleted = request.files.size,
                                    filesTotal = request.files.size,
                                    bytesTransferred = request.totalBytes,
                                    bytesTotal = request.totalBytes,
                                    currentName = "",
                                    currentBytes = 0,
                                    currentSize = 0,
                                    bytesPerSecond = estimator.bytesPerSecond(),
                                    etaSeconds = 0,
                                    state = TransferState.COMPLETED,
                                ),
                            )
                            return request
                        }
                        MessageType.TRANSFER_CANCELLED -> throw ProtocolException(ErrorCode.CANCELLED, "peer cancelled")
                        MessageType.ERROR -> {
                            val error = ProtocolJson.payload<ErrorPayload>(control)
                            throw ProtocolException(ErrorCode.fromWire(error.code), error.code)
                        }
                        else -> throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "unexpected ${control.type}")
                    }
                }
                is Frame.Binary -> throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "binary without FILE_START")
            }
        }
    }

    fun cancel(transferId: String) {
        cancelled.set(true)
        try {
            session.send(
                session.typed(
                    MessageType.TRANSFER_CANCELLED,
                    TransferCancelledPayload(transferId),
                ),
            )
        } catch (_: Exception) {
        }
    }

    private fun sendOne(
        transferId: String,
        file: OutgoingFile,
        index: Int,
        totalFiles: Int,
        overallBefore: Long,
        totalBytes: Long,
        estimator: SpeedEstimator,
        onProgress: (TransferProgress) -> Unit,
    ) {
        session.send(
            session.typed(
                MessageType.FILE_START,
                FileStartPayload(
                    transferId = transferId,
                    fileId = file.meta.fileId,
                    name = file.meta.name,
                    size = file.meta.size,
                    mimeType = file.meta.mimeType,
                    offset = 0,
                ),
            ),
        )
        val digest = IncrementalSha256()
        var sent = 0L
        file.open().use { input ->
            val buffer = ByteArray(ProtocolConstants.CHUNK_SIZE)
            while (true) {
                throwIfCancelled(transferId)
                val n = input.read(buffer)
                if (n < 0) break
                if (n == 0) continue
                val chunk = if (n == buffer.size) buffer else buffer.copyOf(n)
                digest.update(chunk)
                session.sendBinary(UUID.fromString(file.meta.fileId), sent, chunk)
                sent += n
                val overall = overallBefore + sent
                val speed = estimator.onProgress(overall, System.currentTimeMillis())
                onProgress(
                    TransferProgress(
                        transferId = transferId,
                        filesCompleted = index,
                        filesTotal = totalFiles,
                        bytesTransferred = overall,
                        bytesTotal = totalBytes,
                        currentName = file.meta.name,
                        currentBytes = sent,
                        currentSize = file.meta.size,
                        bytesPerSecond = speed,
                        etaSeconds = estimator.etaSeconds(totalBytes - overall),
                        state = TransferState.TRANSFERRING,
                    ),
                )
            }
        }
        if (sent != file.meta.size) {
            throw ProtocolException(ErrorCode.FILE_UNAVAILABLE, "size changed while sending")
        }
        session.send(
            session.typed(
                MessageType.FILE_COMPLETE,
                FileCompletePayload(transferId, file.meta.fileId, sent, digest.hex()),
            ),
        )
    }

    private fun receiveOne(
        request: TransferRequestPayload,
        start: FileStartPayload,
        sinkFactory: ReceiveSinkFactory,
        completed: Int,
        overallBefore: Long,
        estimator: SpeedEstimator,
        onProgress: (TransferProgress) -> Unit,
    ): Long {
        val relative = FilenameConflict.sanitizeRelativePath(start.name)
        val meta = FileMeta(start.fileId, relative.substringAfterLast('/'), start.size, start.mimeType, relative)
        val sink = sinkFactory.open(meta, start.offset)
        val digest = IncrementalSha256()
        try {
            while (sink.bytesWritten < start.size) {
                throwIfCancelled(request.transferId)
                when (val frame = session.receive()) {
                    is Frame.Binary -> {
                        if (!sameFileId(frame.fileId, start.fileId)) {
                            throw ProtocolException(
                                ErrorCode.PROTOCOL_VIOLATION,
                                "unexpected file id start=${start.fileId} frame=${frame.fileId}",
                            )
                        }
                        digest.update(frame.data)
                        sink.write(frame.data)
                        val overall = overallBefore + sink.bytesWritten
                        val speed = estimator.onProgress(overall, System.currentTimeMillis())
                        onProgress(
                            TransferProgress(
                                transferId = request.transferId,
                                filesCompleted = completed,
                                filesTotal = request.files.size,
                                bytesTransferred = overall,
                                bytesTotal = request.totalBytes,
                                currentName = meta.name,
                                currentBytes = sink.bytesWritten,
                                currentSize = start.size,
                                bytesPerSecond = speed,
                                etaSeconds = estimator.etaSeconds(request.totalBytes - overall),
                                state = TransferState.TRANSFERRING,
                            ),
                        )
                    }
                    is Frame.Control -> {
                        val control = ProtocolJson.parse(frame.json)
                        if (control.messageType() == MessageType.FILE_COMPLETE) {
                            val complete = ProtocolJson.payload<FileCompletePayload>(control)
                            if (sink.bytesWritten != complete.bytes) {
                                throw ProtocolException(ErrorCode.CHECKSUM_MISMATCH, "size mismatch")
                            }
                            val actual = digest.hex()
                            if (!Checksums.equalsHex(complete.sha256, actual)) {
                                throw ProtocolException(ErrorCode.CHECKSUM_MISMATCH, "hash mismatch")
                            }
                            sink.commit(complete.sha256)
                            return overallBefore + sink.bytesWritten
                        }
                        if (control.messageType() == MessageType.TRANSFER_CANCELLED) {
                            sink.abort()
                            throw ProtocolException(ErrorCode.CANCELLED, "peer cancelled")
                        }
                        throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "unexpected ${control.type}")
                    }
                }
            }
            val complete = session.receiveControl()
            if (complete.messageType() != MessageType.FILE_COMPLETE) {
                throw ProtocolException(ErrorCode.PROTOCOL_VIOLATION, "expected FILE_COMPLETE")
            }
            val payload = ProtocolJson.payload<FileCompletePayload>(complete)
            if (!Checksums.equalsHex(payload.sha256, digest.hex())) {
                throw ProtocolException(ErrorCode.CHECKSUM_MISMATCH, "hash mismatch")
            }
            sink.commit(payload.sha256)
            return overallBefore + sink.bytesWritten
        } catch (error: Exception) {
            sink.abort()
            throw error
        }
    }

    private fun sameFileId(frameId: UUID, startId: String): Boolean =
        frameId.toString().equals(startId, ignoreCase = true)

    private fun throwIfCancelled(transferId: String) {
        if (cancelled.get()) {
            session.send(
                session.typed(
                    MessageType.TRANSFER_CANCELLED,
                    TransferCancelledPayload(transferId),
                ),
            )
            throw ProtocolException(ErrorCode.CANCELLED, "cancelled")
        }
    }
}
